File: src/Imputer.cpp
#include <map>
#include <sstream>
#include <algorithm>
#include <cmath>
#include "../include/Imputer.h"
#include "../include/common.h"

using namespace std;
using namespace Imputer;


//////////////////// List ////////////////////

template<typename T>
vector<T> List<T>::to_vector() const {
	vector<T>	v;
	const List<T>	*ptr_list = this;
	while(ptr_list != NULL) {
		v.push_back(ptr_list->last);
		ptr_list = ptr_list->first.get();
	}
	
	std::reverse(v.begin(), v.end());
	return v;
}

template<typename T>
typename List<T>::ptrList List<T>::add(
											const ptrList ptr, T val) {
	return ptrList(new List<T>(ptr, val));
}


//////////////////// Region ////////////////////

string Region::str() const {
	stringstream	ss;
	ss << '[' << first << ", " << last << ']';
	return ss.str();
}

vector<tuple<Color, size_t, size_t>> Region::group_by_colors(
														const string& seq) {
	size_t	last = 0U;
	Color	prev_color = -1;
	vector<tuple<Color, size_t, size_t>>	intervals;	// [(Color, first, last)]
	for(auto p = seq.begin(); p != seq.end(); ++p) {
		const Color	color = (Color)(*p) - '0';
		if(prev_color != -1 && color != prev_color) {
			const size_t	j = p - seq.begin();
			intervals.push_back(make_tuple(prev_color, last, j));
			last = p - seq.begin();
		}
		prev_color = color;
	}
	intervals.push_back(make_tuple(prev_color, last, seq.size()));
	return intervals;
}

vector<const Region *> Region::create(const string& seq,
											const vector<double>& cMs) {
	const auto	intervals = group_by_colors(seq);
	
	vector<const Region *>	regions;
	for(auto p = intervals.begin(); p != intervals.end(); ++p) {
		const Color		color = get<0>(*p);
		const size_t	first = get<1>(*p);
		const size_t	last = get<2>(*p);
		const double	cM1 = first == 0 ? cMs[first] :
											(cMs[first-1] + cMs[first]) / 2;
		const double	cM2 = last == seq.size() ? cMs[last-1] :
											(cMs[last-1] + cMs[last]) / 2;
		Region	*region = new Region(color, first, last, cM2 - cM1);
		regions.push_back(region);
	}
	return regions;
}


//////////////////// State ////////////////////

State::State(const std::vector<const Region *>& rs) :
						regions(rs),
						color_list(new List<Color>(regions[0]->get_color())),
						size(1U), num_continuous(1U),
						continuous_length(regions[0]->get_cM()),
						painted_len(0.0) { }

bool State::is_changable(double MIN_CROSSOVER) const {
	if(continuous_length >= MIN_CROSSOVER)
		return true;
	
	// it is painted the same color till now
	ptrColors	colors = color_list;
	while(true) {
		if(colors->back() != last_color())
			return false;
		if(colors->is_null())
			break;
		colors = colors->prev();
	}
	return true;
}

State *State::add(double cM) const {
	const bool	painted = last_color() != next_color();
	ptrColors	colors = List<Color>::add(color_list, last_color());
	const double	new_painted_length = painted_len + (painted ? cM : 0.0);
	return new State(regions, colors, size + 1, num_continuous + 1,
							continuous_length + cM, new_painted_length);
}

State *State::change(double cM) const {
	const Color	color = last_color() == 0 ? 1 : 0;
	const bool	painted = color != next_color();
	ptrColors	colors = List<Color>::add(color_list, color);
	const double	new_painted_length = painted_len + (painted ? cM : 0.0);
	return new State(regions, colors, size + 1, 1, cM, new_painted_length);
}

vector<Color> State::get_colors() const {
	vector<Color>	colors;
	vector<Color>	region_colors = color_list.get()->to_vector();
	for(auto p = region_colors.begin(); p != region_colors.end(); ++p) {
		const size_t	k = p - region_colors.begin();
		const Region	*region = *(regions.begin() + k);
		for(size_t i = 0U; i < region->get_size(); ++i)
			colors.push_back(*p);
	}
	return colors;
}

vector<const State *> State::nexts(double cM, double MIN_CROSSOVER) const {
	vector<const State *>	states;
	if(is_extendable())
		states.push_back(add(cM));
	if(is_changable(MIN_CROSSOVER))
		states.push_back(change(cM));
	return states;
}

ptrState State::min(const vector<ptrState>& states) {
	ptrState	min_state = states.front();
	for(auto p = states.begin() + 1; p != states.end(); ++p) {
		const State	*state = p->get();
		if(state->painted_length() < min_state.get()->painted_length())
			min_state = *p;
	}
	return min_state;
}

// minimize states
vector<ptrState> State::select(const vector<const State *>& states) {
	typedef pair<int,Color>	Key;
	map<Key,vector<ptrState>>	dic;
	for(auto p = states.begin(); p != states.end(); ++p) {
		Key	key((*p)->get_num_continuous(),(*p)->last_color());
		dic[key].push_back(ptrState(*p));
	}
	
	vector<ptrState>	selected_states;
	for(auto p = dic.begin(); p != dic.end(); ++p)
		selected_states.push_back(min(p->second));
	return selected_states;
}

vector<ptrState> State::next(vector<ptrState>& states,
										double cM, double MIN_CROSSOVER) {
	vector<const State *>	new_states;
	for(auto p = states.begin(); p != states.end(); ++p) {
		const auto	next_states = p->get()->nexts(cM, MIN_CROSSOVER);
		new_states.insert(new_states.end(),
							next_states.begin(), next_states.end());
	}
	return select(new_states);
}

string State::str() const {
	stringstream	ss;
	for(auto p = regions.begin(); p != regions.end(); ++p)
		ss << (*p)->str();
	ss << "\n";
	
	const vector<Color>	colors = this->get_colors();
	for(auto p = colors.begin(); p != colors.end(); ++p)
		ss << *p;
	
	return ss.str();
}


//////////////////// Paint ////////////////////

string Imputer::paint(const string& seq, const vector<double>& cMs,
												double MIN_CROSSOVER) {
	const auto	regions = Region::create(seq, cMs);
	if(regions.size() <= 2U) {
		Common::delete_all(regions);
		return seq;
	}
	
	auto	states = vector<ptrState>(1U, ptrState(new State(regions)));
	for(auto p = regions.begin() + 1; p != regions.end(); ++p)
		states = State::next(states, (*p)->get_cM(), MIN_CROSSOVER);
	
	ptrState	state = State::min(states);
	const auto	colors = state.get()->get_colors();
	
	stringstream	ss;
	for(auto p = colors.begin(); p != colors.end(); ++p)
		ss << *p;
	Common::delete_all(regions);
	return ss.str();
}


//////////////////// impute ////////////////////

vector<char> Imputer::create_states(const string& seq) {
	if(seq.find('N') != string::npos) {
		vector<char>	states = { '0', '1', 'N' };
		return states;
	}
	else {
		vector<char>	states = { '0', '1' };
		return states;
	}
}

BaumWelch::TransitionMatrix Imputer::compute_T(double prob) {
	BaumWelch::TransitionMatrix	T;
	for(size_t h0 = 0; h0 < 2; ++h0) {
		for(size_t h1 = 0; h1 < 2; ++h1)
			T[h0*2+h1] = h0 == h1 ? 1.0 - prob : prob;
	}
	return T;
}

string Imputer::impute(const string& seq, const vector<double>& cMs) {
	// Kosambi
	vector<double>	ps;
	for(size_t i = 0; i < cMs.size() - 1; ++i) {
		const double	d = (cMs[i+1] - cMs[i]) / 100;
		const double	r = (exp(d) - exp(-d)) / (exp(d) + exp(-d));
		ps.push_back(r);
	}
	
	vector<BaumWelch::TransitionMatrix>	Ts;
	for(auto p = ps.begin(); p != ps.end(); ++p)
		Ts.push_back(Imputer::compute_T(*p));
	
	const string	hidden_seq = BaumWelch::impute(seq, Ts);
	return hidden_seq;
}

File: python/Imputer.py
from __future__ import annotations

# coding: utf-8
# Imputer.py

from itertools import *
from collections import defaultdict
from math import exp

from typing import Generator, Dict

import Baum_Welch_with_fixed_Ts_log as BWTL
from common import *


#################### Region ####################

class Region:
	def __init__(self, color: str, first: int, last: int, cM: float):
		self.color: str	= color
		self.first: int	= first
		self.last: int	= last
		self.cM: float	= cM
	
	@staticmethod
	def create(seq: str, cMs: list[float]) -> list[Region]:
		regions = []
		last = 0
		for color, w in groupby(seq):
			n = sum(1 for _ in w)
			first, last = last, last + n
			cM1 = cMs[0] if first == 0 else (cMs[first-1] + cMs[first]) / 2
			cM2 = cMs[-1] if last == len(seq) else (cMs[last-1] + cMs[last]) / 2
			regions.append(Region(color, first, last, cM2 - cM1))
		return regions


#################### State ####################

class State:
	def __init__(self, regions: list[Region], colors, size: int,
						n: int, length: float, p_len: float, min_c: float):
		self.regions: list[Region] = regions
		# 効率化のためのリスト構造
		# typingでうまく型を表せない
		self.new_colors = colors	# [[[...], str], str]
		self.size: int = size
		self.num_continuous: int = n
		self.continuous_length: float = length
		self.painted_length: float = p_len
		self.MIN_CROSSOVER: float = min_c
	
	def last_color(self) -> str:
		return self.new_colors[1]
	
	def to_list(self) -> list[str]:
		cs = self.new_colors
		v: list[str] = []
		while cs:
			v = cs[1:] + v
			cs = cs[0]
		return v
	
	def extend(self, cM: float) -> State:
		painted = self.last_color() != self.next_color()
		return State(self.regions,
					 [self.new_colors, self.last_color()],
					 self.size + 1,
					 self.num_continuous + 1,
					 self.continuous_length + cM,
					 self.painted_length + (cM if painted else 0.0),
					 self.MIN_CROSSOVER)
	
	def change(self, cM: float) -> State:
		color = '1' if self.last_color() == '0' else '0'
		painted = color != self.next_color()
		return State(self.regions,
					 [self.new_colors, color],
					 self.size + 1,
					 1,
					 cM,
					 self.painted_length + (cM if painted else 0.0),
					 self.MIN_CROSSOVER)
	
	def current_color(self) -> str:
		return self.regions[self.size-1].color
	
	def next_color(self) -> str:
		return self.regions[self.size].color
	
	def is_extendable(self) -> bool:
		return (self.size != len(self.regions) - 1 or
				self.next_color() == self.last_color())
	
	def rev_colors(self) -> Generator[str, None, None]:
		new_colors = self.new_colors
		while new_colors:
			yield new_colors[1]
			new_colors = new_colors[0]
	
	def is_changable(self) -> bool:
		color = self.next_color()
		return (self.continuous_length >= self.MIN_CROSSOVER or
				# it is painted the same color till now
				all(c == self.last_color() for c in self.rev_colors()))
	
	def nexts(self, cM: float) -> Generator[State, None, None]:
		if self.is_extendable():
			yield self.extend(cM)
		if self.is_changable():
			yield self.change(cM)
	
	@staticmethod
	def init(regions: list[Region], min_c: float) -> State:
		new_colors = [[], regions[0].color]
		return State(regions, new_colors, 1, 1, regions[0].cM, 0.0, min_c)


#################### paint ####################

def paint(seq: str, cMs: list[float], min_c: float) -> str:
	def select(states: list[State]) -> list[State]:
		dic: Dict[tuple[int, str], list[State]] = defaultdict(list)
		for state in states:
			dic[(state.num_continuous, state.last_color())].append(state)
		
		return [ min(ss, key=lambda s: s.painted_length)
									for ss in dic.values() ]
	
	regions = Region.create(seq, cMs)
	if len(regions) <= 2:
		return seq
	
	# DP
	states: list[State] = [State.init(regions, min_c)]
	for region in regions[1:]:
		new_states = [ s2 for s1 in states for s2 in s1.nexts(region.cM) ]
		states = select(new_states)
	
	state = min(states, key=lambda s: s.painted_length)
	new_colors = list(state.rev_colors())[::-1]
	return ''.join(color * (region.last - region.first)
						for region, color in zip(regions, new_colors))


#################### impute ####################

def is_all_same_without_N(seq: str) -> bool:
	prev_c = '-'
	for c in seq:
		if c == 'N':
			continue
		elif prev_c == '-':
			prev_c = c
		elif c != prev_c:
			return False
	else:
		return True

def create_same_color_string(seq: str, default_color: str):
	if all(c == 'N' for c in seq):
		c_base = default_color
	else:
		c_base = next(c for c in seq if c != 'N')
	return c_base * len(seq)

def compute_T(p: float, hidden_states: list[str]
							) -> Dict[tuple[str, str], float]:
	return { (h1, h2): 1.0-p if h1 == h2 else p
					for h1, h2 in product(hidden_states, repeat=2) }

def impute(seq: str, hidden_states: list[str],
					states: list[str], cMs: list[float]) -> str:
	ds = [ (p2-p1)/100 for p1, p2 in zip(cMs, cMs[1:]) ]
	ps = [ (exp(d) - exp(-d)) / (exp(d) + exp(-d)) for d in ds ]
	Ts = [ compute_T(p, hidden_states) for p in ps ]
	pi, E = BWTL.Baum_Welch(Ts, seq, states, hidden_states)
	A = BWTL.reverse_probs(E, hidden_states, states)
	hidden_seq = BWTL.Viterbi(seq, states, hidden_states, pi, Ts, A)
	return hidden_seq


#################### main ####################

__all__ = ['paint', 'impute', 'is_all_same_without_N',
								'create_same_color_string']
